casper/src/main/resources/PoS.rhox

```diff
@@ -81,25 +105,39 @@ new  PoS,
             } |
 
             contract PoS (@"pay", @amount, returnCh) = {
-              new vaultCh, transferCh, userCh in {
+              new vaultCh, transferCh, userCh, processPayCh,
+                  depositCh,
+                  computeSum, totalStakeCh,
+                  computeDelta, rewardsDeltaCh,
+                  computeMergeMap, newRewardsCh in {
                 getUser!(*userCh) |
 
-                for(@userPk <- userCh) {
-                  deposit!(userPk, amount, *returnCh)
+                runMVar!(*stateCh, *processPayCh, *returnCh) |
+                for(@(rewards, bonds), payResultCh <- processPayCh;
+                    @userPk <- userCh) {
+                  deposit!(userPk, amount, *depositCh) |
+
+                  distributeRewards!(amount, bonds, rewards, *newRewardsCh) |
+
+                  for(@depositResult <- depositCh;
+                      @newRewards <- newRewardsCh) {
+                    payResultCh!((newRewards, bonds), depositResult)
+                  }
                 }
               }
             } |
 
             contract PoS(@"slash", @blockHash, return) = {
-              new userCh, invalidBlocksCh, getInvalidBlocks(`rho:casper:invalidBlocks`) in {
+              new userCh, invalidBlocksCh,
+                  getInvalidBlocks(`rho:casper:invalidBlocks`) in {
                 getInvalidBlocks!(*invalidBlocksCh) |
                 getUser!(*userCh) |
-                for (@invalidBlocks <- invalidBlocksCh; @bonds <- bondsCh; @userPk <- userCh) {
+                for (@invalidBlocks <- invalidBlocksCh; @(rewards, bonds) <- stateCh; @userPk <- userCh) {
                   new toBeSlashed in {
                     toBeSlashed!(invalidBlocks.getOrElse(blockHash, userPk)) |
                     for (@validator <- toBeSlashed) {
                       // TODO: Transfer to coop wallet instead of just simply setting bonds to 0
```

Comment:
shouldn't that TODO be a JIRA issue?

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3584